### ISSUE TITLE
[CI] DocsClientYamlTestSuiteIT timed out after 30 minutes

### ISSUE THREAD

**Issue by andreidan** [elastic/elasticsearch#49753]
Build failure https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+matrix-java-periodic/ES_BUILD_JAVA=openjdk12,ES_RUNTIME_JAVA=corretto11,nodes=general-purpose/362/console

```
Suite timeout exceeded (>= 1800000 msec)
```

Build scan available at https://gradle-enterprise.elastic.co/s/lp2vd4vqmj25o

**Comment by mark-vieira** [comment]
This is happening increasingly often over the past couple of days due to the scope of the suite simply slowly increasing over time. I think we may need to bump this timeout in the interim until we sort out a better solution like splitting it up.

**Comment by mark-vieira** [comment]
Now I am indeed starting to believe we've introduced some kind of regression. This is failing consistently across many OSes and JREs.

![image](https://user-images.githubusercontent.com/4106672/72373100-70a78600-36bc-11ea-90d1-c5ffedcea4d1.png)

**Comment by mark-vieira** [comment]
So something has clearly changed to make this test suite suddenly run so much slower. The reason I don't believe we are seeing this failure in intake and pull request checks is that those are split up into multiple smaller parallel builds and therefore this is much less contention on the worker then for the full periodic builds.

There's no single test that's running long. It simply looks as though all test cases are just a bit slower, adding up to a pretty substantial runtime hit.

Another interesting data point is this is only happening in `master` and `7.x`. We are not seeing the increased runtimes in `7.5`. My closest guess at a potential culprit for a large change that might impact performance here is perhaps https://github.com/elastic/elasticsearch/pull/50907.

@ywelsch @DaveCTurner Any thoughts? Any chance that change might have an adverse effect on performance on a system with lot of other CPU and IO contention (lots of other concurrent tests).

@danielmitterdorfer Have you seen any recent changes in performance benchmarks that might indicate some kind regression might have been introduced?

**Comment by hendrikmuhs** [comment]
For me this look like a duplicate of #49579

**Comment by DaveCTurner** [comment]
I suspect https://github.com/elastic/elasticsearch/pull/50907 and https://github.com/elastic/elasticsearch/pull/50928. Although these changes save potentially hundreds of `fsync` calls from some cluster state updates in large clusters it will add an `fsync` or two to cluster state updates in tiny clusters, and these tests use a tiny cluster which does a lot of cluster state updates.

Do these tests run using a ramdisk? If not, can we move them to use a ramdisk? That would make `fsync` calls free again.

**Comment by hendrikmuhs** [comment]
While investigations are running which hopefully result in a proper fix I will increase the timeout as a temporary solution, because to many builds are failing at the moment.

**Comment by hendrikmuhs** [comment]
I increased the timeout to 35 minutes. **This is a temporary solution, please revert once we have a proper solution**

**Comment by tvernum** [comment]
It doesn't look like 35 minutes is enough. Suite is still timing out.
- [7.6 fedora-29 2020-01-20](https://gradle-enterprise.elastic.co/s/74wh7k3wqdyuk/console-log?anchor=10090)
- [7.x fedora-29 2020-01-20](https://gradle-enterprise.elastic.co/s/tb3r2ya5hc3be/console-log?anchor=10673)

I'll push the timeout up further.

**Comment by danielmitterdorfer** [comment]
> @danielmitterdorfer Have you seen any recent changes in performance benchmarks that might indicate some kind regression might have been introduced?

Our [benchmarks](https://elasticsearch-benchmarks.elastic.co/#) don't indicate any major regressions recently but the workload and cluster configuration is quite different here so [David's suggestion](https://github.com/elastic/elasticsearch/issues/49753#issuecomment-575522038) seems reasonable.

**Comment by mark-vieira** [comment]
> Do these tests run using a ramdisk? If not, can we move them to use a ramdisk? That would make fsync calls free again.

Yes, this is timing out both in builds that use a ramdisk and those that don't.

**Comment by mark-vieira** [comment]
This suite is still timing out even at 40 minutes. Its strange that this doesn't seem to be affecting (at least not to a perceptible degree) other test suites in a significant way. My only guess here is that something shifted in terms of the execution graph of the build and now there is simply more contention on the system when running this very long test suite. This also explains why we aren't seeing the timeouts in intake or pull request builds, because they are cut up into smaller pieces and there is less intrabuild contention.

I'm curious if we bump this to something very large, say 1 hour, if the total build test execution time ends up being about the same. If so, tells me we are doing the same amount of work, but perhaps the docs tests taking longer means other tests executing more quickly. Right now it's hard to do apples to apples comparisons across builds because we are failing early due to this timeout.

**Comment by astefan** [comment]
There is one suite where both DocsClientYamlTestSuiteIT and XPackRestIT timed out.


https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.6+matrix-java-periodic/ES_BUILD_JAVA=openjdk13,ES_RUNTIME_JAVA=corretto11,nodes=general-purpose/14/console
[7.5.3] https://gradle-enterprise.elastic.co/s/6zkf5kjcqwvcc
[6.8.7] https://gradle-enterprise.elastic.co/s/poka253asglpc
https://gradle-enterprise.elastic.co/s/aaqsqtdunbilk

```
org.elasticsearch.smoketest.DocsClientYamlTestSuiteIT classMethod
org.elasticsearch.xpack.test.rest.XPackRestIT classMethod
org.elasticsearch.xpack.test.rest.XPackRestIT test {p0=privileges/10_basic/Test put and delete privileges}
org.elasticsearch.smoketest.DocsClientYamlTestSuiteIT test {yaml=reference/aggregations/bucket/daterange-aggregation/line_78}
```

**Comment by astefan** [comment]
Another XPackRestIT timeout.


https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=centos-6&&immutable/528/console
[7.5.3] https://gradle-enterprise.elastic.co/s/yfraoimkm2oty
[7.7.0] https://gradle-enterprise.elastic.co/s/o5gnkc5kbkhv4 
[7.6.0] https://gradle-enterprise.elastic.co/s/q5nocksdi7wak 
[7.5.3] https://gradle-enterprise.elastic.co/s/nrvafb443xw4k 
[7.6.0] https://gradle-enterprise.elastic.co/s/rorimvt3wpcsi 
[7.7.0] https://gradle-enterprise.elastic.co/s/zexvtet3swsw2
https://gradle-enterprise.elastic.co/s/blpyy3dggqnqg

**Comment by mark-vieira** [comment]
Overall build times after increasing the timeouts have increased vs a week ago before we had these timeout issues which to me seems to imply a general slowness in test execution that has been introduced somehow.

You can see in the graph below of intake build times that there's a clear bump starting last week.
![image](https://user-images.githubusercontent.com/4106672/72918990-7b38cf80-3cfb-11ea-99d3-6b35bac7d05f.png)

It's unlikely this is infrastructure related since this doesn't affect the `7.5` branch and we get consistent timeouts across operating systems, java runtime versions, etc. There's also no related build changes during that time. The only major change was the upgrade to Gradle 6.1 which was after we started seeing these timeouts.

**Comment by mark-vieira** [comment]
As another datapoint, we specifically **do not** use a ramdisk in our unix compatibility tests due to wanting to exercise the distro FS implementations. We do however on the Java compatibility tests, which are also still timing out.

**Comment by polyfractal** [comment]
Master docs yaml tests timed out today.  I couldn't find evidence of any particular test killing the cluster or the test itself failing, so looks to be another instance of general slowdown.

https://gradle-enterprise.elastic.co/s/xrrdvuh3wza2s

**Comment by ywelsch** [comment]
@mark-vieira can you establish a base-line performance for `docs:check` on our CI machines?

On my machine, for example, it takes 32 minutes, running on an SSD, but on RamDisk, it takes 16 minutes, which brings it well below the range of the timeout.

**Comment by polyfractal** [comment]
Another maybe-useful data point: I just saw three timeouts come in back-to-back, and they were all from the same host.  Perhaps there is something specific to certain hosts/machines/OS combinations that is exacerbating the issue

**Comment by mark-vieira** [comment]
>  I just saw three timeouts come in back-to-back, and they were all from the same host

We use ephemeral workers so we never run multiple builds on a single agent. Perhaps the same _type_ of host. Our [compatibility matrix tests](https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility) show this timing out across all operating systems so I think specific host types is not the issue. Also, as I mentioned before, this regression is not in `7.5` so that indicates to me something that got introduced in `7.x` and later.

**Comment by mark-vieira** [comment]
> On my machine, for example, it takes 32 minutes, running on an SSD, but on RamDisk, it takes 16 minutes, which brings it well below the range of the timeout.

Keep in mind this is running _only_ a single test suite so resource contention is very low. This is the same reason why we generally don't see timeouts in pull-request or intake builds because those jobs are split up into smaller pieces with fewer concurrently executing tests.

The graph below is average execution time for this suite over the past 30 days for intake/PR builds (these use a ramdisk). You can see while we are still well below the timeout threshold there is a clear bump in execution time around the time of Jan 12-13. The additional load of our periodic tests (which run the entire test suite in a single build) is enough to push this over the edge.

![image](https://user-images.githubusercontent.com/4106672/73009182-e8646780-3dc4-11ea-8d69-f4bc7be5b9dd.png)

**Comment by mark-vieira** [comment]
Here's the same graph but for `XPackRestIT`. The bump in execution time isn't as dramatic but it's there, and the timetables align.

![image](https://user-images.githubusercontent.com/4106672/73009394-56109380-3dc5-11ea-8daf-b3c2805dc55b.png)

**Comment by mark-vieira** [comment]
Here are the graphs for `DocsClientYamlTestSuiteIT` per branch across all CI builds:

**7.5**
![image](https://user-images.githubusercontent.com/4106672/73009916-30d05500-3dc6-11ea-8edc-584c2907784f.png)

**7.x**
![image](https://user-images.githubusercontent.com/4106672/73009955-42b1f800-3dc6-11ea-9ddc-d36c31c1769c.png)

**master**
![image](https://user-images.githubusercontent.com/4106672/73009994-59584f00-3dc6-11ea-8826-408f364df689.png)

The clear disparity between branches makes me want to rule out anything infrastructure related given that all this stuff runs on the same infra. That said, there _might_ have been some change in `7.x` and later that interacts awkwardly with our infra, or some infra change in some way.

**Comment by mark-vieira** [comment]
Here's an interesting thing I just found. Here's the syslogs for the CI agent running two identical build configurations:

**7.5:**https://infra-stats.elastic.co/app/infra#/logs?_g=()&flyoutOptions=(flyoutId:!n,flyoutVisibility:hidden,surroundingLogsId:!n)&logFilter=(expression:'host.name:elasticsearch-ci-immutable-ubuntu-1604-1579811827669019501',kind:kuery)&logPosition=(position:(tiebreaker:82329295,time:1579813230000),streamLive:!f)&logTextview=(textScale:small,wrap:!f)
**7.x:**https://infra-stats.elastic.co/app/infra#/logs?_g=()&flyoutOptions=(flyoutId:!n,flyoutVisibility:hidden,surroundingLogsId:!n)&logFilter=(expression:'host.name:elasticsearch-ci-immutable-fedora-29-1579709304051438758',kind:kuery)&logPosition=(position:(tiebreaker:53366760,time:1579713794000),streamLive:!f)&logTextview=(textScale:small,wrap:!t)

The 7.x builds are producing 50x more output in the syslogs due to what seems to be us logging every syscall. This is undoubtably contributing to some overhead, not only to perform the logging but whatever instrumentation is enabled that's performing this logging. Anyone have any ideas why we might be getting all this extra logging all of a sudden?

**Comment by mark-vieira** [comment]
Ok, I take back the conclusion in my previous comment. The additional syscall logging seems to be specific to certain OSes, not branches, and we have extended test execution times even on systems where this logging isn't present so this doesn't seem to be the cause of our problem. 
Damn, back to the drawing board.

**Comment by ywelsch** [comment]
I've found the source for slowness here and will open a PR shortly. The gist here is that with ILM/SLM we do a lot of unnecessary setup / teardown work on each test. Compounded with the slightly slower cluster state storage mechanism, this causes the tests to run much slower.

On the core/infra side, we could also look at speeding up plugin installs. It currently takes 30-40 seconds at the beginning of `docs:integTest` to just install the plugins:

```
[2020-01-24T15:22:56.594355Z] [BUILD] Installing 16 plugins
[2020-01-24T15:23:41.084037Z] [BUILD] Starting Elasticsearch process
```

**Comment by rjernst** [comment]
> On the core/infra side, we could also look at speeding up plugin installs.

While we could and should (now that we can install them all in one plugin command), the timeout here is in the test suite, while we do plugin installation externally from gradle.

**Comment by mark-vieira** [comment]
I don't believe the solution proposed in #51418 is going to help here. As I've stated before, we already run most of our CI builds in ramdisk workspace and those builds continue to timeout at 40 minutes whereas we were running under 20 minutes before.

**Comment by mark-vieira** [comment]
This looks to have done the trick. Thanks @ywelsch!

**master**
![image](https://user-images.githubusercontent.com/4106672/73394822-2c5adf00-4293-11ea-942a-88dabb9015a3.png)

**7.x**
![image](https://user-images.githubusercontent.com/4106672/73394856-3e3c8200-4293-11ea-96bc-3b5625b0b321.png)

**Comment by ywelsch** [comment]
We got another timeout on elastic+elasticsearch+master+multijob-darwin-compatibility: https://gradle-enterprise.elastic.co/s/fczylflkgrebm

Looking at the output, the tests are running at an excruciatingly slow pace (3-4 seconds per test). Perhaps darwin tests are not using RAMDisk, or doing too many things in parallel?

**Comment by mark-vieira** [comment]
Thanks for reporting this @ywelsch. This looks to be an issue with that macos Jenkins worker. We've seen a number of build failures specific to this machine for which I have an open infra issue for (https://github.com/elastic/infra/issues/17621).

You can see this test is passing just fine and running in a reasonable time on other mac workers.
![image](https://user-images.githubusercontent.com/4106672/73880198-d0540580-4812-11ea-8af9-159dd8481671.png)

Since this is an infra problem, and I've reached out regarding this, I'm going to close this issue again.
